scripts/zephyr_module.py

```diff
@@ -215,20 +220,52 @@ def main():
     cmake = ""
     sanitycheck = ""
 
+    dep_modules = []
+    sorted_modules = []
+    start_modules = []
+
     for project in projects:
         # Avoid including Zephyr base project as module.
         if project == os.environ.get('ZEPHYR_BASE'):
             continue
 
         meta = process_module(project)
         if meta:
-            kconfig += process_kconfig(project, meta)
-            cmake += process_cmake(project, meta)
-            sanitycheck += process_sanitycheck(project, meta)
+            section = meta.get('build', dict())
+            deps = section.get('depends', [])
+            if not deps:
+                start_modules.append( [project, meta] )
+            else:
+                dep_modules.append( [project, meta, deps] )
         elif project in extra_modules:
             sys.exit(f'{project}, given in ZEPHYR_EXTRA_MODULES, '
                      'is not a valid zephyr module')
 
+    # This will do a topology search to ensure the modules are ordered
+    # according to dependency settings.
+    while start_modules:
+        # We want to preserve order as close to that returned by west.
+        # Thus remove elements from beginning, instead of pop().
+        node = start_modules[0]
+        del start_modules[0]
+        sorted_modules.append(node)
+        node_name = PurePath(node[0]).name
+        for module in dep_modules:
+            if node_name in module[2]:
+                module[2].remove(node_name)
+                if not module[2]:
+                    start_modules.append(module)
+                    dep_modules.remove(module)
+
+    if dep_modules:
```

Comment:
Could you please document or explain more clearly what happens when a dependency is missing? I expected to see an error somewhere and it's not popping out at me.

Reply:
This is no a dependency missing, but if there is still entries in `dep_modules` and processing of `start_modules` has been completed, it means that the remaining modules has a cyclic dependency.

Adding a dependency is not required, but if a module is dependent on another module, that module must be earlier in the sorted list.

Reply:
added some comments.

Reply:
Is it possible to separate the cases of cyclic dependencies versus unmet dependencies? I think unmet dependencies should be relatively easy to filter out. The rest would be cyclic. I think the output would be more readable if the different cases produced different output personally, but I won't NAK over it.

Reply:
Now printing:
```
CMake Error at /projects/github/ncs/zephyr/cmake/zephyr_module.cmake:45 (message):
  Unmet or cyclic dependencies in modules:

  /projects/github/ncs/modules/hal/cmsis depends on: ['bda_splelnig']

  /projects/github/ncs/modules/fs/fatfs depends on: ['nordic']

  /projects/github/ncs/modules/hal/nordic depends on: ['fatfs']
```
don't like two split the two, cause there could be both unmet deps AND cyclic deps at the same time.

But now the deps are printed a long with the modules, which should make it very easy to find the root cause of the failing dependency.

Reply:
Perfect. Thanks.